firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/UpdateAppClient.java

```diff
@@ -79,10 +109,185 @@ private void redirectToPlayForAabUpdate(String downloadUrl, Activity currentActi
+          updateProgressOnMainThread(totalSize, downloadedSize, UpdateStatus.DOWNLOADING);
+        }
+      }
+      // completion
+      updateProgressOnMainThread(totalSize, downloadedSize, UpdateStatus.DOWNLOADED);
+
+    } catch (IOException e) {
+      setDownloadTaskCompletionError(
+          new FirebaseAppDistributionException(
+              Constants.ErrorMessages.NETWORK_ERROR,
+              FirebaseAppDistributionException.Status.DOWNLOAD_FAILURE));
+    }
+
+    try {
+      // check that file is actual JAR
+      new JarFile(apkFile);
+    } catch (Exception e) {
+      setDownloadTaskCompletionError(
+          new FirebaseAppDistributionException(
+              Constants.ErrorMessages.NETWORK_ERROR,
+              FirebaseAppDistributionException.Status.DOWNLOAD_FAILURE));
+    }
+
+    downloadTaskCompletionSource.setResult(apkFile);
+  }
+
+  private File getApkFileForApp(String fileName) {
+    return new File(firebaseApp.getApplicationContext().getFilesDir(), fileName);
+  }
+
+  private String getApplicationName() {
+    try {
+      Context context = firebaseApp.getApplicationContext();
+      return context.getApplicationInfo().loadLabel(context.getPackageManager()).toString();
+    } catch (Exception e) {
+      Log.e(TAG, "Unable to retrieve App name");
+      return "";
+    }
+  }
+
+  private void setDownloadTaskCompletionError(FirebaseAppDistributionException e) {
+    if (downloadTaskCompletionSource != null
+        && !downloadTaskCompletionSource.getTask().isComplete()) {
+      downloadTaskCompletionSource.setException(e);
+    }
+  }
+
+  public Task<Void> install(String path, Activity currentActivity) {
+    Intent intent = new Intent(currentActivity, InstallActivity.class);
+    intent.putExtra("INSTALL_PATH", path);
+    this.installCancellationTokenSource = new CancellationTokenSource();
+    this.installTaskCompletionSource =
+        new TaskCompletionSource<>(installCancellationTokenSource.getToken());
+    InstallActivity.registerOnCompletionListener(this.installTaskCompletionSource);
+    intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+    currentActivity.startActivity(intent);
+    return installTaskCompletionSource.getTask();
+  }
+
   private void setUpdateAppTaskCompletionError(FirebaseAppDistributionException e) {
     if (updateAppTaskCompletionSource != null
         && !updateAppTaskCompletionSource.getTask().isComplete()) {
       updateAppTaskCompletionSource.setException(e);
     }
   }
+
+  private void setUpdateAppErrorWithDefault(
+      Exception e, FirebaseAppDistributionException defaultFirebaseException) {
+    if (e instanceof FirebaseAppDistributionException) {
+      setUpdateAppTaskCompletionError((FirebaseAppDistributionException) e);
+    } else {
+      setUpdateAppTaskCompletionError(defaultFirebaseException);
+    }
+  }
+
+  private void updateProgressOnMainThread(
+      long totalBytes, long downloadedBytes, UpdateStatus status) {
+    downloadHandler.post(
```

Comment:
I am unfamiliar with this post method. Could you explain it to me?

Reply:
From what I've read post is basically analogous to thread.start(), using the thread that the handler is attached to which in this case is the main/UI thread. So essentially it 'posts' the runnable on the main thread's message queue